### ISSUE TITLE
What's the meaning of "sparse gradient has high density"?

### ISSUE THREAD

**Issue by ChaiBapchya** [horovod/horovod#2906]
Refer : https://horovod.readthedocs.io/en/stable/api.html#horovod.tensorflow.DistributedGradientTape

https://github.com/horovod/horovod/blob/cbb343dec8d54bc2634acf11ac295a551519905b/horovod/tensorflow/keras/__init__.py#L74-L77

Sparse gradient will have lots of 0s and sparse non-zero values.
What makes that sparse gradient tensor "highly-dense"? If its highly dense then its no longer sparse right?

**Comment by tgaddair** [comment]
Hey @ChaiBapchya, a tensor can be represented using a "sparse representation" (i.e., a map) while still being relatively dense (meaning, there are few missing values). When we say `sparse_as_dense`, we mean "convert this sparse representation / map into a dense representation / vector when performing the allreduce".

**Comment by ChaiBapchya** [comment]
Thanks!

**Comment by naila81** [comment]
how sparse gradient problems are appropriate for ADAM method
